discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -759,6 +758,82 @@ impl FeedbackScope for SqliteStorage {
         tx.commit().await?;
         Ok(document)
     }
+
+    async fn update_time_spend(
+        &self,
+        document: document::Id,
+        view_mode: ViewMode,
+        duration: Duration,
+    ) -> Result<TimeSpendDocumentView, Error> {
+        const MS_PER_DAY: i64 = 1000 * 60 * 60 * 24 * 7;
+        // If someone somehow manages to overflow the sqlite integer it turns into a real,
+        // which is a problem. As precaution we cap the added view time to the already
+        // extremely high value of one day.
+        let view_time_ms = i64::try_from(duration.as_millis())
+            .ok()
+            .unwrap_or(MS_PER_DAY);
+        let view_time_ms = min(view_time_ms, MS_PER_DAY);
+
+        let mut tx = self.pool.begin().await?;
+
+        sqlx::query(
+            "INSERT INTO ViewTimes(documentId, viewMode, viewTimeMs) VALUES (?, ?, ?)
+                ON CONFLICT DO UPDATE SET viewTimeMs =  viewTimeMs + excluded.viewTimeMs;",
+        )
+        .bind(document)
+        .bind(view_mode as u32)
+        .bind(view_time_ms)
+        .execute(&mut tx)
+        .await?;
+
+        let view = sqlx::query_as::<_, QueryTimeSpendDocumentView>(
+            "SELECT
+                em.embedding,
+                sum(vt.viewTimeMs) as aggregatedViewTime,
+                ur.userReaction
+            FROM Embedding          AS em
+            JOIN ViewTimes          AS vt   USING (documentId)
+            LEFT JOIN UserReaction  AS ur   USING (documentId)
+            WHERE documentId = ?;",
+        )
+        .bind(document)
+        .fetch_one(&mut tx)
+        .await
+        .on_row_not_found(Error::NoDocument(document))
+        .and_then(TryInto::try_into)?;
+
+        tx.commit().await?;
+
+        Ok(view)
+    }
+}
+
+#[derive(FromRow)]
+#[sqlx(rename_all = "camelCase")]
+struct QueryTimeSpendDocumentView {
+    embedding: Vec<u8>,
+    aggregated_view_time: i64,
+    user_reaction: Option<u32>,
+}
+
+impl TryFrom<QueryTimeSpendDocumentView> for TimeSpendDocumentView {
+    type Error = Error;
+
+    fn try_from(
+        QueryTimeSpendDocumentView {
+            embedding,
+            aggregated_view_time,
+            user_reaction,
+        }: QueryTimeSpendDocumentView,
+    ) -> Result<Self, Self::Error> {
+        Ok(TimeSpendDocumentView {
+            smbert_embedding: embedding.try_into()?,
+            last_reaction: user_reaction_from_db(user_reaction)?,
+            aggregated_view_time: Duration::from_millis(
+                u64::try_from(aggregated_view_time).unwrap_or(0),
```

Comment:
❓ should we make a distinction here and map negative values to `0` and too large values to `u64::MAX`/`MS_PER_DAY`?

Reply:
:rocket: For this use case `u32` is enough so I

- limited the view time to u32::MAX 
  - both when getting it from the app
  - and when upsert increasing it in SQL
- use `u32` in struct

:pushpin: sql not supporting unsigned integers is annoying, in different places we use different strategies to work around it. In all cases the out of bound handling falls into the  "extremely unlikely edge cases which likely can only appear when some bugs are involved" category. At the same time we don't really want our app to ever lock up, because due to some bug (from us, or team blue, or system time glitch) there are some bad stats in the system we might not even care too much about. I have to think about it a bit more, maybe we should add some `u63` type, idk. :thinking: